### PR TITLE
fixed: File times for vfs addons (fixes #15426)

### LINKED ISSUE #15426: Sort by date doesn't work in sftp add-on (vfs issue)
### Describe the bug
Sort by date doesn't work in the sftp add-on. When you browse an sftp location and choose sort by date, it instead sorts by filename.

## Possible Fix
I looked through the code and found that the sftp add-on is populating the timestamp of the file:
`kodi::vfs::CDirEntry entry;
[...]
entry.SetDateTime(attributes->mtime64);`

However, in VFSEntry.cpp (in kodi core), you find this code:
`static void VFSDirEntriesToCFileItemList(int num_entries,
                                         VFSDirEntry* entries,
                                         CFileItemList& items)
{
  for (int i=0;i<num_entries;++i)
  {
    CFileItemPtr item(new CFileItem());
    item->SetLabel(entries[i].label);
    item->SetPath(entries[i].path);
    item->m_dwSize = entries[i].size;
    //item->m_dateTime = entries[i].mtime;`

Note the commented out line. I don't know why this was done. I believe m_dateTime is a CDateTime which has an override for assigning a time_t value so fixing this issue should be as simple as uncommenting that line.

### To Reproduce
1. Add an sftp source
2. Browse files, select your source
3. Choose to sort by date

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version: 18

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by arnova**
This fixes the missing file times in eg. the sftp vfs addon which also causes other problems like the videoscanner not working properly. Note that the sftp addon also requires a small fix to make this work, which will follow shortly.

### CODE REVIEW

**Review by notspiff — APPROVED** [review]

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/kodi_vfs_types.h [hunk 1]
```diff
@@ -41,7 +41,7 @@ extern "C"
     char* path;              //!< item path
     int num_props;           //!< Number of properties attached to item
     VFSProperty* properties; //!< Properties
-    //FILETIME mtime;          //!< Mtime for file represented by item
+    time_t date_time;        //!< Mtime for file represented by item
```
Comment (arnova) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/kodi_vfs_types.h]: @notspiff : Are you fine with the change from FILETIME to time_t here? It's obviously a change of the interface.

Comment (notspiff) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/kodi_vfs_types.h]: Not really, since there was no interface before.

**Review by Rechi — CHANGES_REQUESTED** [review]
`ADDON_(GLOBAL|INSTANCE)_VERSION_(FILESYSTEM|VFS)(_MIN)?` has to be adjusted at the following places.
https://github.com/xbmc/xbmc/blob/792be7bfd45ba43508a5a21f42b3e515e0336b59/xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h#L58-L59
https://github.com/xbmc/xbmc/blob/792be7bfd45ba43508a5a21f42b3e515e0336b59/xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h#L114-L115

**Comment by arnova** [comment]
> `ADDON_(GLOBAL|INSTANCE)_VERSION_(FILESYSTEM|VFS)(_MIN)?` has to be adjusted at the following places.
> https://github.com/xbmc/xbmc/blob/792be7bfd45ba43508a5a21f42b3e515e0336b59/xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h#L58-L59
> 
> 
> https://github.com/xbmc/xbmc/blob/792be7bfd45ba43508a5a21f42b3e515e0336b59/xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h#L114-L115

Just bump the minor numbers, I assume?

**Comment by Rechi** [comment]
You adjusted the patch version, not minor.

**Comment by arnova** [comment]
> You adjusted the patch version, not minor.

@Rechi : Ok, so instead of bumping patch version, minor should be updated in this case?

**Comment by Rechi** [comment]
You asked to bump minor version not patch version, that's why I added the above comment.

For `ADDON_INSTANCE_VERSION_VFS` I would bump minor version as it break ABI.
I'm not sure if anything breaks because of the Filesystem changes, so there a patch bump could be enough.

Also keep in mind this PR can't be backported in the current state as it would break ABI compatibility.

**Comment by arnova** [comment]
> You asked to bump minor version not patch version, that's why I added the above comment.
> 
> For `ADDON_INSTANCE_VERSION_VFS` I would bump minor version as it break ABI.
> I'm not sure if anything breaks because of the Filesystem changes, so there a patch bump could be enough.
> 
> Also keep in mind this PR can't be backported in the current state as it would break ABI compatibility.

@Rechi : Done

**Review by AlwinEsch — APPROVED** [review]
Looks good and makes more sense to make the DateTime usable.

**Comment by AlwinEsch** [comment]
Maybe only edit the commit name to and include the version change to them:
`[addons] fixed Missing bits for file times in vfs addons`

The `[...]` on begin makes it easier to look over a commit history

**Comment by arnova** [comment]
@AlwinEsch : Updated. Ok now?

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/kodi_vfs_types.h (hunk 1 above)
Comment (garbear) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/kodi_vfs_types.h]: Maybe include the clock in the comment, like "in Unix time"

**Comment by AlwinEsch** [comment]
@arnova >  Updated. Ok now? 

:confused::grin: what do you think, I see no difference to this morning?

**Comment by arnova** [comment]
> @arnova > Updated. Ok now?
> 
>  how do you mean? I see no difference to this morning.



> @arnova > Updated. Ok now?
> 
>  how do you mean? I see no difference to this morning.

@AlwinEsch : Should be *really* updated now ;-)

**Comment by arnova** [comment]
> Also keep in mind this PR can't be backported in the current state as it would break ABI compatibility.

@Rechi : I just gave this another thought, but I think there's no problem in backporting this since only the internals were adjusted, there are no new or renamed functions. So unless I'm missing something, I think we can backport this to Leia.

**Comment by arnova** [comment]
@notspiff : Added a small fix for an inconsistency in the type used for num_props. And do you see any problems backporting this to Leia?

**Comment by arnova** [comment]
Good to go?

**Comment by AlwinEsch** [comment]
From my side yes, can go in.

Only found following addons where need version increase (by quick search):
- audiodecoder.2sf
- audiodecoder.asap
- audiodecoder.dumb
- audiodecoder.fluidsynth
- audiodecoder.gme
- audiodecoder.gsf
- audiodecoder.ssf
- audiodecoder.modplug
- audiodecoder.ncsf
- audiodecoder.nosefart
- audiodecoder.openmpt
- audiodecoder.organya
- audiodecoder.sidplay
- audiodecoder.snesapu
- audiodecoder.stsound
- audiodecoder.timidity
- audiodecoder.qsf
- audiodecoder.upse
- audiodecoder.usf
- audiodecoder.vgmstream
- audiodecoder.wsr
- inputstream.adaptive
- peripheral.joystick
- pvr.vuplus
- pvr.zattoo
- screensaver.asterwave (already request present)
- screensavers.rsxs (already request present)
- vfs.libarchive
- vfs.nfs
- vfs.rar
- vfs.sacd
- vfs.sftp
- vfs.smb2
- visualization.milkdrop
- visualization.spectrum

**Comment by AlwinEsch** [comment]
I make a request for this addons

**Comment by AlwinEsch** [comment]
You was a bit fast with merge :sweat_smile:

Normally better to have the addon requests mostly ready, so the time between Kodi and addon change is more less.

**Comment by arnova** [comment]
> You was a bit fast with merge 😅
> 
> Normally better to have the addon requests mostly ready, so the time between Kodi and addon change is more less.

Ok, thanks for the heads up

**Pull request merged**